File: java/core/src/org/daisy/urakawa/property/xml/XmlAttribute.java
package org.daisy.urakawa.property.xml;

import java.net.URI;

import org.daisy.urakawa.WithPresentation;
import org.daisy.urakawa.event.DataModelChangedEvent;
import org.daisy.urakawa.event.Event;
import org.daisy.urakawa.event.EventHandler;
import org.daisy.urakawa.event.IEventHandler;
import org.daisy.urakawa.event.IEventListener;
import org.daisy.urakawa.event.property.xml.ValueChangedEvent;
import org.daisy.urakawa.exception.IsAlreadyInitializedException;
import org.daisy.urakawa.exception.IsNotInitializedException;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.nativeapi.IXmlDataReader;
import org.daisy.urakawa.nativeapi.IXmlDataWriter;
import org.daisy.urakawa.progress.IProgressHandler;
import org.daisy.urakawa.progress.ProgressCancelledException;
import org.daisy.urakawa.xuk.XukDeserializationFailedException;
import org.daisy.urakawa.xuk.XukSerializationFailedException;

/**
 * Reference implementation of the interface.
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 */
public class XmlAttribute extends WithPresentation implements IXmlAttribute {
	IXmlProperty mParent;
	String mLocalName;
	String mNamespaceUri;
	String mValue;
	protected IEventHandler<Event> mDataModelEventNotifier = new EventHandler();
	protected IEventHandler<Event> mValueChangedEventNotifier = new EventHandler();

	public <K extends DataModelChangedEvent> void notifyListeners(K event)
			throws MethodParameterIsNullException {
		if (event == null) {
			throw new MethodParameterIsNullException();
		}
		if (ValueChangedEvent.class.isAssignableFrom(event.getClass())) {
			mValueChangedEventNotifier.notifyListeners(event);
		}
		mDataModelEventNotifier.notifyListeners(event);
	}

	public <K extends DataModelChangedEvent> void registerListener(
			IEventListener<K> listener, Class<K> klass)
			throws MethodParameterIsNullException {
		if (klass == null || listener == null) {
			throw new MethodParameterIsNullException();
		}
		if (ValueChangedEvent.class.isAssignableFrom(klass)) {
			mValueChangedEventNotifier.registerListener(listener, klass);
		} else {
			mDataModelEventNotifier.registerListener(listener, klass);
		}
	}

	public <K extends DataModelChangedEvent> void unregisterListener(
			IEventListener<K> listener, Class<K> klass)
			throws MethodParameterIsNullException {
		if (klass == null || listener == null) {
			throw new MethodParameterIsNullException();
		}
		if (ValueChangedEvent.class.isAssignableFrom(klass)) {
			mValueChangedEventNotifier.unregisterListener(listener, klass);
		} else {
			mDataModelEventNotifier.unregisterListener(listener, klass);
		}
	}

	/**
	 * 
	 */
	public XmlAttribute() {
		mParent = null;
		mLocalName = null;
		mNamespaceUri = "";
		mValue = "";
	}

	public IXmlAttribute copy() {

		IXmlAttribute attr = new XmlAttribute();
		try {
			attr.setPresentation(getPresentation());
			attr.setLocalName(getLocalName());
			attr.setNamespace(getNamespace());
			attr.setValue(getValue());
		} catch (IsAlreadyInitializedException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (IsNotInitializedException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (MethodParameterIsEmptyStringException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (MethodParameterIsNullException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		return attr;
	}

	public String getValue() {
		return mValue;
	}

	public void setValue(String newValue)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		if (newValue == null) {
			throw new MethodParameterIsNullException();
		}
		if (newValue.length() == 0) {
			throw new MethodParameterIsEmptyStringException();
		}
		String prevVal = mValue;
		mValue = newValue;
		if (mValue != prevVal) {
			notifyListeners(new ValueChangedEvent(this, mValue, prevVal));
		}
	}

	public String getLocalName() throws IsNotInitializedException {
		if (mLocalName == null || mLocalName.length() == 0) {
			throw new IsNotInitializedException();
		}
		return mLocalName;
	}

	public String getNamespace() throws IsNotInitializedException {
		if (mNamespaceUri == null) {
			throw new IsNotInitializedException();
		}
		return mNamespaceUri;
	}

	public void setQName(String localname, String namespace)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		if (localname == null || namespace == null) {
			throw new MethodParameterIsNullException();
		}
		if (localname.length() == 0) {
			throw new MethodParameterIsEmptyStringException();
		}
		if (localname != mLocalName || namespace != mNamespaceUri) {
			IXmlProperty parent = getParent();
			if (parent != null) {
				try {
					parent.removeAttribute(this);
				} catch (XmlAttributeDoesNotExistException e) {
					// Should never happen
					throw new RuntimeException("WTF ??!", e);
				}
			}
			mLocalName = localname;
			mNamespaceUri = namespace;
			if (parent != null) {
				parent.setAttribute(this);
			}
		}
	}

	public void setLocalName(String newName)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException {
		if (newName == null) {
			throw new MethodParameterIsNullException();
		}
		if (newName.length() == 0) {
			throw new MethodParameterIsEmptyStringException();
		}
		if (newName != mLocalName) {
			IXmlProperty parent = getParent();
			if (parent != null) {
				try {
					parent.removeAttribute(this);
				} catch (XmlAttributeDoesNotExistException e) {
					// Should never happen
					throw new RuntimeException("WTF ??!", e);
				}
			}
			mLocalName = newName;
			if (parent != null) {
				parent.setAttribute(this);
			}
		}
	}

	public void setNamespace(String newNS)
			throws MethodParameterIsNullException {
		if (newNS == null) {
			throw new MethodParameterIsNullException();
		}
		if (newNS != mNamespaceUri) {
			IXmlProperty parent = getParent();
			if (parent != null) {
				try {
					parent.removeAttribute(this);
				} catch (XmlAttributeDoesNotExistException e) {
					// Should never happen
					throw new RuntimeException("WTF ??!", e);
				}
			}
			mNamespaceUri = newNS;
			if (parent != null) {
				parent.setAttribute(this);
			}
		}
	}

	public IXmlProperty getParent() {
		return mParent;
	}

	public void setParent(IXmlProperty prop) {
		mParent = prop;
	}

	@Override
	protected void clear() {
		if (getParent() != null) {
			try {
				getParent().removeAttribute(this);
			} catch (MethodParameterIsNullException e) {
				// Should never happen
				throw new RuntimeException("WTF ??!", e);
			} catch (XmlAttributeDoesNotExistException e) {
				// Should never happen
				throw new RuntimeException("WTF ??!", e);
			}
		}
		mLocalName = null;
		mNamespaceUri = "";
		mValue = "";
		// super.clear();
	}

	@Override
	protected void xukInAttributes(IXmlDataReader source, IProgressHandler ph)
			throws MethodParameterIsNullException,
			XukDeserializationFailedException, ProgressCancelledException {
		if (source == null) {
			throw new MethodParameterIsNullException();
		}
		// To avoid event notification overhead, we bypass this:
		if (false && ph != null && ph.notifyProgress()) {
			throw new ProgressCancelledException();
		}
		String name = source.getAttribute("localName");
		if (name == null || name.length() == 0) {
			throw new XukDeserializationFailedException();
		}
		String ns = source.getAttribute("namespaceUri");
		if (ns == null)
			ns = "";
		try {
			setLocalName(name);
		} catch (MethodParameterIsEmptyStringException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		setNamespace(ns);
	}

	@Override
	protected void xukOutAttributes(IXmlDataWriter destination, URI baseUri,
			IProgressHandler ph) throws MethodParameterIsNullException,
			XukSerializationFailedException, ProgressCancelledException {
		if (destination == null || baseUri == null) {
			throw new MethodParameterIsNullException();
		}
		if (ph != null && ph.notifyProgress()) {
			throw new ProgressCancelledException();
		}
		if (mLocalName == null || mLocalName.length() == 0) {
			throw new XukSerializationFailedException();
		}
		destination.writeAttributeString("localName", mLocalName);
		if (mNamespaceUri != "")
			destination.writeAttributeString("namespaceUri", mNamespaceUri);
		// super.xukOutAttributes(destination, baseUri, ph);
	}

	@Override
	public String toString() {
		String displayName = mLocalName == null ? "null" : mLocalName;
		try {
			if (getNamespace() != "")
				displayName = getNamespace() + ":" + displayName;
		} catch (IsNotInitializedException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		return String.format("{1}: {2}='{3}'", super.toString(), displayName,
				getValue().replace("'", "''"));
	}

	@SuppressWarnings("unused")
	@Override
	protected void xukOutChildren(IXmlDataWriter destination, URI baseUri,
			IProgressHandler ph) throws XukSerializationFailedException,
			MethodParameterIsNullException, ProgressCancelledException {
		/**
		 * Does nothing.
		 */
	}
}

File: branches/1.1_before_major_refactor_2.0/csharp/core/property/xml/XmlAttribute.cs
using System;
using System.Xml;
using urakawa.property;
using urakawa.xuk;

namespace urakawa.property.xml
{
	/// <summary>
	/// Default implementation of <see cref="XmlAttribute"/>
	/// </summary>
	public class XmlAttribute : WithPresentation
	{
		internal class ValueChangedEventArgs : urakawa.events.DataModelChangedEventArgs
		{
			public ValueChangedEventArgs(XmlAttribute src, string newVal, string prevVal) : base(src)
			{
				SourceXmlAttribute = src;
				NewValue = newVal;
				PreviousValue = prevVal;
			}
			public readonly XmlAttribute SourceXmlAttribute;
			public readonly string NewValue;
			public readonly string PreviousValue;
		}

		internal event EventHandler<ValueChangedEventArgs> valueChanged;
		private void notifyValueChanged(XmlAttribute src, string newVal, string prevVal)
		{
			EventHandler<ValueChangedEventArgs> d = valueChanged;
			if (d != null) d(this, new ValueChangedEventArgs(src, newVal, prevVal));
		}

		XmlProperty mParent;
		string mLocalName;
		string mNamespaceUri;
		string mValue;

		/// <summary>
		/// Constructor setting the parent <see cref="XmlProperty"/>
		/// </summary>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when the parent is <c>null</c>
		/// </exception>
		protected internal XmlAttribute()
		{
			mParent = null;
			mLocalName = null;
			mNamespaceUri = "";
			mValue = "";
		}

		#region XmlAttribute Members
		/// <summary>
    /// Creates a copy of the <see cref="XmlAttribute"/>
    /// </summary>
		/// <param name="newParent">The parent xml property of the copy</param>
    /// <returns>The copy</returns>
		/// <exception cref="exception.FactoryCannotCreateTypeException">
		/// Thrown when the <see cref="IGenericPropertyFactory"/> of the <see cref="Presentation"/> 
		/// to which <c>this</c> belongs is not a subclass of <see cref="IXmlPropertyFactory"/>
		/// </exception>
    public XmlAttribute copy(XmlProperty newParent)
		{
			return export(getParent().getPresentation(), newParent);
		}

		/// <summary>
		/// Exports the xml attribute to a given destination presentation 
		/// with a given parent <see cref="XmlProperty"/>
		/// </summary>
		/// <param name="destPres">The given destination presentation</param>
		/// <param name="parent">The given parent xml property</param>
		/// <returns>The exported xml attribute</returns>
		public XmlAttribute export(Presentation destPres, XmlProperty parent)
		{
			if (destPres == null)
			{
				throw new exception.MethodParameterIsNullException(
					"The destination Presentation can not be null");
			}
			if (parent == null)
			{
				throw new exception.MethodParameterIsNullException(
					"The parent XmlProperty can not be null");
			}
			if (parent.getPresentation() != destPres)
			{
				throw new exception.OperationNotValidException(
					"The parent XmlProperty must belong to the destination Presentation");
			}
			string xukLN = getXukLocalName();
			string xukNS = getXukNamespaceUri();
			XmlAttribute exportAttr = destPres.getPropertyFactory().createXmlAttribute(xukLN, xukNS);
			if (exportAttr == null)
			{
				throw new exception.FactoryCannotCreateTypeException(String.Format(
					"The xml property factory does not support creating xml attributes matching QName {0}:{1}", 
					xukLN, xukNS));
			}
			exportAttr.setQName(getLocalName(), getNamespaceUri());
			exportAttr.setValue(getValue());
			return exportAttr;
		}



    /// <summary>
    /// Gets the value of gthe <see cref="XmlAttribute"/>
    /// </summary>
    /// <returns>The value</returns>
    public string getValue()
		{
			return mValue;
		}

    /// <summary>
    /// Sets the value of the <see cref="XmlAttribute"/>
    /// </summary>
    /// <param name="newValue">The new value</param>
    public void setValue(string newValue)
		{
			string prevVal = mValue;
			mValue = newValue;
			if (mValue != prevVal)
			{
				notifyValueChanged(this, mValue, prevVal);
			}
		}

    /// <summary>
    /// Gets the namespace of the <see cref="XmlAttribute"/>
    /// </summary>
    /// <returns>The namespace</returns>
    public string getNamespaceUri()
		{
				return mNamespaceUri;
		}

    /// <summary>
    /// Gets the local localName of the <see cref="XmlAttribute"/>
    /// </summary>
    /// <returns>The local localName</returns>
    public string getLocalName()
		{
			if (mLocalName == null)
			{
				throw new exception.IsNotInitializedException(
					"The XmlAttribute has not been initialized with a local name");
			}
			return mLocalName;
		}

    /// <summary>
    /// Sets the QName of the <see cref="XmlAttribute"/> 
    /// </summary>
    /// <param name="newNamespaceUri">The namespace part of the new QName</param>
    /// <param name="newLocalName">The localName part of the new QName</param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Throw when <paramref name="newNamespaceUri"/> or <paramref name="newLocalName"/> is <c>null</c>
		/// </exception>
		/// <exception cref="exception.MethodParameterIsEmptyStringException">
		/// Thrown when <paramref name="newLocalName"/> is an <see cref="String.Empty"/>
		/// </exception>
		/// <remarks>
		/// If the <see cref="XmlAttribute"/> has already been set on a <see cref="XmlProperty"/>,
		/// setting the QName will overwrite any <see cref="XmlAttribute"/> of the owning <see cref="XmlProperty"/>
		/// with matching QName
		/// </remarks>
		public void setQName(string newLocalName, string newNamespaceUri)
		{
			if (newLocalName == null)
			{
				throw new exception.MethodParameterIsNullException("The local localName must not be null");
			}
			if (newLocalName == String.Empty)
			{
				throw new exception.MethodParameterIsEmptyStringException("The local localName must not be empty");
			}
			if (newNamespaceUri == null)
			{
				throw new exception.MethodParameterIsNullException("The namespace uri must not be null");
			}
			if (newLocalName != mLocalName || newNamespaceUri != mNamespaceUri)
			{
				XmlProperty parent = getParent();
				if (parent != null)
				{
					parent.removeAttribute(this);
				}
				mLocalName = newLocalName;
				mNamespaceUri = newNamespaceUri;
				if (parent != null)
				{
					parent.setAttribute(this);
				}
			}
		}

    /// <summary>
    /// Gets the parent <see cref="XmlProperty"/> of <c>this</c>
    /// </summary>
    /// <returns></returns>
    public XmlProperty getParent()
		{
			return mParent;
		}

		/// <summary>
		/// Sets the parent <see cref="XmlProperty"/> of <c>this</c>. 
		/// Is intended for internal use by the owning <see cref="XmlProperty"/>,
		/// calling this method may lead to corruption of the data model
		/// </summary>
		/// <param name="newParent">The new parent</param>
		public void setParent(XmlProperty newParent)
		{
			mParent = newParent;
		}
		#endregion

		#region IXUKAble members 

		/// <summary>
		/// Clears the <see cref="XmlAttribute"/> data
		/// </summary>
		protected override void clear()
		{
			if (getParent() != null)
			{
				getParent().removeAttribute(this);
			}
			mLocalName = null;
			mNamespaceUri = "";
			mValue = "";
			base.clear();
		}

		/// <summary>
		/// Reads the attributes of a XmlAttribute xuk element.
		/// </summary>
		/// <param name="source">The source <see cref="XmlReader"/></param>
		protected override void xukInAttributes(XmlReader source)
		{
			string name = source.GetAttribute("localName");
			if (name == null || name == "")
			{
				throw new exception.XukException("LocalName attribute of XmlAttribute element is missing");
			}
            string value = source.GetAttribute("Value");
            if (value == null)
            {
                throw new exception.XukException("Value attribute of XmlAttribute element is missing");
            }
            mValue = value;

			string ns = source.GetAttribute("namespaceUri");
			if (ns == null) ns = "";
			setQName(name, ns);
		}

		/// <summary>
		/// Writes the attributes of a XmlAttribute element
		/// </summary>
		/// <param name="destination">The destination <see cref="XmlWriter"/></param>
		/// <returns>A <see cref="bool"/> indicating if the write was succesful</returns>
		/// <param name="baseUri">
		/// The base <see cref="Uri"/> used to make written <see cref="Uri"/>s relative, 
		/// if <c>null</c> absolute <see cref="Uri"/>s are written
		/// </param>
		protected override void xukOutAttributes(XmlWriter destination, Uri baseUri)
		{
			//localName is required
			if (mLocalName == "")
			{
				throw new exception.XukException("The XmlAttribute has no name");
			}
			destination.WriteAttributeString("localName", mLocalName);

            if (mValue == null)
            {
                throw new exception.XukException("The XmlAttribute has no value");
            }
            destination.WriteAttributeString("Value", mValue);

			if (mNamespaceUri != "") destination.WriteAttributeString("namespaceUri", mNamespaceUri);
			base.xukOutAttributes(destination, baseUri);
		}
		#endregion

		/// <summary>
		/// Gets a <see cref="string"/> representation of the attribute
		/// </summary>
		/// <returns>The <see cref="string"/> representation</returns>
		public override string ToString()
		{
			string displayName = mLocalName==null?"null":mLocalName;
			if (getNamespaceUri() != "") displayName = getNamespaceUri() + ":" + displayName;
			return String.Format("{1}: {2}='{3}'", base.ToString(), displayName, getValue().Replace("'", "''"));
		}

	}
}
